src/lib.rs

```diff
@@ -469,14 +469,39 @@ impl<U: AsRef<[u8]>> CheckBase32<()> for U {
     }
 }
 
-#[derive(Clone, Copy, PartialEq, Eq)]
-enum Case {
+/// Letter case.
+///
+/// From [BIP-173]: Uppercase/lowercase
+///
+/// > The lowercase form is used when determining a character's value for checksum purposes.
+/// >
+/// > Encoders MUST always output an all lowercase Bech32 string. If an uppercase version of the
+/// > encoding result is desired, (e.g.- for presentation purposes, or QR code use), then an
+/// > uppercasing procedure can be performed external to the encoding process.
+/// >
+/// > Decoders MUST NOT accept strings where some characters are uppercase and some are lowercase (such strings are referred to as mixed case strings).
+/// >
+/// > For presentation, lowercase is usually preferable, but inside QR codes uppercase SHOULD be
+/// > used, as those permit the use of alphanumeric mode, which is 45% more compact than the normal
+/// > byte mode.
+///
+/// [BIP173]: <https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#user-content-Bech32>
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Case {
+    /// Uppercase letters.
     Upper,
+    /// Lowercase letters.
     Lower,
+    /// No letter case, i.e., all characters were digits.
     None,
```

Comment:
I don't like this variant and think it should be folded into `Lower`.  AFAIK there is no HRP defined by anybody which is all digits, so any codepaths that involve this variant will be unused.

I think that `check_hrp` should just return `()` or an error. If people want to check case-consistency of the whole string they should use `check_characters` and check this at the same time that they're checking that every non-HRP character is a valid bech32 string.

Maaybe it's marginally more efficient to do things this way, but adding a whole extra type to the API, which is an enum where 1/3 of the variants will literally never be used, is a pretty big cost.